core/src/main/java/io/grpc/internal/SharingExecutorService.java

```diff
@@ -0,0 +1,114 @@
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Wraps an existing executor, that keeps track of only the tasks it has added.
+ */
+public final class SharingExecutorService extends AbstractExecutorService {
+  private final Map<CleanupRunnable, Future<?>> activeTasks =
+      new HashMap<CleanupRunnable, Future<?>>();
+
+  private final ExecutorService delegate;
+  private boolean shutdown;
+  private boolean terminated;
+
+  public SharingExecutorService(ExecutorService delegate) {
+    this.delegate = delegate;
+  }
+
+  @Override
+  public synchronized void shutdown() {
+    shutdown = true;
+    if (activeTasks.isEmpty()) {
+      terminated = true;
+      notifyAll();
+    }
+  }
+
+  @Override
+  public synchronized List<Runnable> shutdownNow() {
+    shutdown();
+    List<Runnable> neverRan = new ArrayList<Runnable>(activeTasks.size());
+    for (Entry<CleanupRunnable, Future<?>> entry : activeTasks.entrySet()) {
+      // Since the underlying executor may have started running this, this is best effort.
+      boolean success = entry.getValue().cancel(true);
+      if (success) {
+        neverRan.add(entry.getKey().r);
+      }
+    }
+    return neverRan;
+  }
+
+  @Override
+  public synchronized boolean isShutdown() {
+    return shutdown;
+  }
+
+  @Override
+  public synchronized boolean isTerminated() {
+    return terminated;
+  }
+
+  @Override
+  public synchronized boolean awaitTermination(long timeout, TimeUnit unit)
+      throws InterruptedException {
+    long remainingNanos = unit.toNanos(timeout);
+    long deadline = System.nanoTime() + remainingNanos;
+    while (!terminated && remainingNanos > 0) {
+      unit.timedWait(this, remainingNanos);
+      remainingNanos = deadline - System.nanoTime();
+    }
+    return terminated;
+  }
+
+  @Override
+  public synchronized void execute(final Runnable command) {
+    checkState(!shutdown, "Executor shutting down");
+    CleanupRunnable cleanupRunnable = new CleanupRunnable(command);
+    Future<?> f = delegate.submit(cleanupRunnable);
```

Comment:
The put() operation on activeTasks is preconditioned shutdown being false.  Making it volatile could still result  in new tasks being added if execute() is called, then shutdown(), and ending with execute() completing.